src/Common/OptimizedRegularExpression.cpp

```diff
@@ -262,32 +356,77 @@ void OptimizedRegularExpressionImpl<thread_safe>::analyze(
+                required_substring.prefix = candidate_it->second == 0;
+                required_substring.suffix = candidate_it + 1 == trivial_substrings.end();
             }
         }
     }
     else if (!trivial_substrings.empty())
     {
-        required_substring = trivial_substrings.front().first;
-        required_substring_is_prefix = trivial_substrings.front().second == 0;
+        required_substring.literal = trivial_substrings.front().first;
+        required_substring.prefix = trivial_substrings.front().second == 0;
+        required_substring.suffix = true;
+    }
+
+    /// if it is xxx|xxx|xxx, we should call the next xxx|xxx recursively and collect the result.
+    if (has_alternative_on_depth_0)
+    {
+        /// compare the quality of required substring and alternatives and choose the better one.
+        if (shortest_alter_length(global_alters) < required_substring.literal.size())
+            global_alters = {required_substring};
+        Literals next_alternatives;
+        /// this two vals are useless, xxx|xxx cannot be trivial nor prefix.
+        bool next_is_trivial = true;
+        pos = analyzeImpl(regexp, pos, required_substring, next_is_trivial, next_alternatives);
+        /// For xxx|xxx|xxx, we only conbine the alternatives and return a empty required_substring.
+        if (next_alternatives.empty() || shortest_alter_length(next_alternatives) < required_substring.literal.size())
+        {
+            global_alters.push_back(required_substring);
+        }
+        else
+        {
+            global_alters.insert(global_alters.end(), next_alternatives.begin(), next_alternatives.end());
+        }
+        required_substring.clear();
     }
 
+    return pos;
+
 /*    std::cerr
         << "regexp: " << regexp
         << ", is_trivial: " << is_trivial
         << ", required_substring: " << required_substring
         << ", required_substring_is_prefix: " << required_substring_is_prefix
         << std::endl;*/
 }
+}
 
+template <bool thread_safe>
+void OptimizedRegularExpressionImpl<thread_safe>::analyze(
+        std::string_view regexp_,
+        std::string & required_substring,
+        bool & is_trivial,
+        bool & required_substring_is_prefix,
+        std::vector<std::string> & alternatives)
+{
+    Literals alter_literals;
+    Literal required_lit;
+    analyzeImpl(regexp_, regexp_.data(), required_lit, is_trivial, alter_literals);
+    required_substring = std::move(required_lit.literal);
+    required_substring_is_prefix = required_lit.prefix;
+    for (auto & lit : alter_literals)
+        alternatives.push_back(std::move(lit.literal));
+}
 
 template <bool thread_safe>
 OptimizedRegularExpressionImpl<thread_safe>::OptimizedRegularExpressionImpl(const std::string & regexp_, int options)
 {
-    analyze(regexp_, required_substring, is_trivial, required_substring_is_prefix);
+    std::vector<std::string> alternatives; /// this vector collects patterns in (xx|xx|xx). for now it's not used.
```

Comment:
To make things a bit more clear, you might rename `alternatives` to `alternativesDummy`.

Reply:
There are some users of OptimizedRegularExpression outside the regexp tree dictionary (pretty much all SQL functions for matching regular expressions). Do they become slower since the logic now collects "alternatives"? If yes, consider making two constructors - one of the will collect alternatives, the other won't.